File: static/js/audio-enhancements.js
// Audio Enhancements for Web - DSP processing using Web Audio API
// Matches Android implementation: bandpass filter, AGC, pre-emphasis, noise gate, click removal

class AudioEnhancementsProcessor extends AudioWorkletProcessor {
    constructor() {
        super();

        // Sample rate (will be set by AudioContext, typically 48000)
        this.sampleRate = sampleRate;

        // Bandpass filter state (80 Hz - 7000 Hz)
        this.initBandpassFilter();

        // AGC state
        this.agcCurrentGain = 1.0;
        this.agcTarget = 0.25; // Target RMS (25% of max)
        this.agcAttack = 0.001;
        this.agcRelease = 0.0001;
        this.agcMaxGain = 8.0;
        this.agcMinGain = 0.5;

        // Pre-emphasis state
        this.preEmphasisPrevSample = 0;
        this.preEmphasisCoeff = 0.97;

        // Noise gate state
        this.gateIsOpen = false;
        this.gateOpenThreshold = 0.02;  // Normalized (0-1)
        this.gateCloseThreshold = 0.01;
        this.gateFadeCounter = 0;
        this.gateAttackSamples = 10;
        this.gateReleaseSamples = 50;

        // Click removal state (median filter)
        this.clickWindowSize = 5;
        this.clickWindow = new Float32Array(this.clickWindowSize);
        this.clickWindowIndex = 0;
    }

    initBandpassFilter() {
        const lowCutoff = 80.0;
        const highCutoff = 7000.0;

        // High-pass filter coefficients (removes < 80 Hz)
        const omega_hp = 2.0 * Math.PI * lowCutoff / this.sampleRate;
        const sn_hp = Math.sin(omega_hp);
        const cs_hp = Math.cos(omega_hp);
        const alpha_hp = sn_hp / (2.0 * Math.sqrt(2.0));

        const hp_b0 = (1.0 + cs_hp) / 2.0;
        const hp_b1 = -(1.0 + cs_hp);
        const hp_b2 = (1.0 + cs_hp) / 2.0;
        const hp_a0 = 1.0 + alpha_hp;
        const hp_a1 = -2.0 * cs_hp;
        const hp_a2 = 1.0 - alpha_hp;

        this.hp_a0 = hp_b0 / hp_a0;
        this.hp_a1 = hp_b1 / hp_a0;
        this.hp_a2 = hp_b2 / hp_a0;
        this.hp_b1 = hp_a1 / hp_a0;
        this.hp_b2 = hp_a2 / hp_a0;

        // Low-pass filter coefficients (removes > 7000 Hz)
        const omega_lp = 2.0 * Math.PI * highCutoff / this.sampleRate;
        const sn_lp = Math.sin(omega_lp);
        const cs_lp = Math.cos(omega_lp);
        const alpha_lp = sn_lp / (2.0 * Math.sqrt(2.0));

        const lp_b0 = (1.0 - cs_lp) / 2.0;
        const lp_b1 = 1.0 - cs_lp;
        const lp_b2 = (1.0 - cs_lp) / 2.0;
        const lp_a0 = 1.0 + alpha_lp;
        const lp_a1 = -2.0 * cs_lp;
        const lp_a2 = 1.0 - alpha_lp;

        this.lp_a0 = lp_b0 / lp_a0;
        this.lp_a1 = lp_b1 / lp_a0;
        this.lp_a2 = lp_b2 / lp_a0;
        this.lp_b1 = lp_a1 / lp_a0;
        this.lp_b2 = lp_a2 / lp_a0;

        // Filter state
        this.hp_x1 = 0; this.hp_x2 = 0;
        this.hp_y1 = 0; this.hp_y2 = 0;
        this.lp_x1 = 0; this.lp_x2 = 0;
        this.lp_y1 = 0; this.lp_y2 = 0;
    }

    process(inputs, outputs) {
        const input = inputs[0];
        const output = outputs[0];

        if (!input || !input[0]) return true;

        const inputChannel = input[0];
        const outputChannel = output[0];
        const blockSize = inputChannel.length;

        // Process each sample
        for (let i = 0; i < blockSize; i++) {
            let sample = inputChannel[i];

            // 1. Bandpass filter (80 Hz - 7000 Hz)
            sample = this.applyBandpassFilter(sample);

            // 2. Click/pop removal
            sample = this.removeClick(sample);

            // 3. Noise gate
            const gateMultiplier = this.applyNoiseGate(sample);
            sample *= gateMultiplier;

            // 4. Pre-emphasis (if gate is open)
            if (gateMultiplier > 0.1) {
                sample = this.applyPreEmphasis(sample);
            }

            outputChannel[i] = sample;
        }

        // 5. AGC on the whole block (after other processing)
        this.applyAGC(outputChannel, blockSize);

        return true;
    }

    applyBandpassFilter(sample) {
        // High-pass filter
        const hp_y = this.hp_a0 * sample + this.hp_a1 * this.hp_x1 + this.hp_a2 * this.hp_x2
                    - this.hp_b1 * this.hp_y1 - this.hp_b2 * this.hp_y2;
        this.hp_x2 = this.hp_x1;
        this.hp_x1 = sample;
        this.hp_y2 = this.hp_y1;
        this.hp_y1 = hp_y;

        // Low-pass filter (creates bandpass)
        const lp_y = this.lp_a0 * hp_y + this.lp_a1 * this.lp_x1 + this.lp_a2 * this.lp_x2
                    - this.lp_b1 * this.lp_y1 - this.lp_b2 * this.lp_y2;
        this.lp_x2 = this.lp_x1;
        this.lp_x1 = hp_y;
        this.lp_y2 = this.lp_y1;
        this.lp_y1 = lp_y;

        return lp_y;
    }

    removeClick(sample) {
        // Add to circular buffer
        this.clickWindow[this.clickWindowIndex] = sample;
        this.clickWindowIndex = (this.clickWindowIndex + 1) % this.clickWindowSize;

        // Calculate median
        const sorted = Array.from(this.clickWindow).sort((a, b) => a - b);
        const median = sorted[Math.floor(this.clickWindowSize / 2)];

        // Detect spike
        const diff = Math.abs(sample - median);
        const threshold = 0.3; // Normalized

        return diff > threshold ? median : sample;
    }

    applyNoiseGate(sample) {
        const amplitude = Math.abs(sample);

        // Hysteresis
        const shouldBeOpen = this.gateIsOpen
            ? amplitude > this.gateCloseThreshold
            : amplitude > this.gateOpenThreshold;

        // Update gate state
        if (shouldBeOpen && !this.gateIsOpen) {
            this.gateIsOpen = true;
            this.gateFadeCounter = 0;
        } else if (!shouldBeOpen && this.gateIsOpen) {
            this.gateIsOpen = false;
            this.gateFadeCounter = this.gateReleaseSamples;
        }

        // Calculate fade multiplier
        let multiplier = 1.0;
        if (!this.gateIsOpen || this.gateFadeCounter > 0) {
            if (this.gateIsOpen) {
                // Fading in
                multiplier = Math.min(1.0, this.gateFadeCounter / this.gateAttackSamples);
                if (this.gateFadeCounter < this.gateAttackSamples) {
                    this.gateFadeCounter++;
                }
            } else {
                // Fading out or closed
                multiplier = this.gateFadeCounter > 0
                    ? this.gateFadeCounter / this.gateReleaseSamples
                    : 0.0;
                if (this.gateFadeCounter > 0) {
                    this.gateFadeCounter--;
                }
            }
        }

        return multiplier;
    }

    applyPreEmphasis(sample) {
        const emphasized = sample - this.preEmphasisCoeff * this.preEmphasisPrevSample;
        this.preEmphasisPrevSample = sample;
        return Math.max(-1.0, Math.min(1.0, emphasized));
    }

    applyAGC(buffer, length) {
        // Calculate RMS
        let sumSquares = 0;
        for (let i = 0; i < length; i++) {
            sumSquares += buffer[i] * buffer[i];
        }
        const rms = Math.sqrt(sumSquares / length);

        // Calculate desired gain
        const desiredGain = rms > 0 ? this.agcTarget / rms : 1.0;

        // Smooth gain changes
        if (desiredGain < this.agcCurrentGain) {
            // Attack
            this.agcCurrentGain += (desiredGain - this.agcCurrentGain) * this.agcAttack;
        } else {
            // Release
            this.agcCurrentGain += (desiredGain - this.agcCurrentGain) * this.agcRelease;
        }

        // Clamp gain
        this.agcCurrentGain = Math.max(this.agcMinGain,
                                       Math.min(this.agcMaxGain, this.agcCurrentGain));

        // Apply gain
        for (let i = 0; i < length; i++) {
            buffer[i] = Math.max(-1.0, Math.min(1.0, buffer[i] * this.agcCurrentGain));
        }
    }
}

registerProcessor('audio-enhancements-processor', AudioEnhancementsProcessor);

File: android/app/src/main/java/com/euno/app/audio/AudioEnhancements.kt
package com.euno.app.audio

import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min

/**
 * Additional audio enhancements for improved speech quality.
 * Includes AGC, pre-emphasis, noise gate, and click removal.
 *
 * Processing pipeline:
 * 1. Click/pop removal - Remove impulse noise
 * 2. Noise gate - Suppress background noise
 * 3. Pre-emphasis - Boost high frequencies for clarity
 * 4. AGC - Normalize volume levels
 */
class AudioEnhancements {

    // ==================== Automatic Gain Control (AGC) ====================

    /**
     * Automatic Gain Control - normalizes audio volume for consistent levels.
     *
     * Adapts to speaker volume and microphone distance, ensuring:
     * - Quiet speakers are boosted
     * - Loud speakers don't clip
     * - Consistent transcription accuracy
     *
     * @param buffer 16-bit PCM buffer to process in-place
     * @param length Number of bytes to process
     */
    private var agcTarget = 8000.0       // Target RMS level (~25% of max)
    private var agcCurrentGain = 1.0     // Current gain multiplier
    private val agcAttack = 0.001        // How fast to reduce gain (prevent clipping)
    private val agcRelease = 0.0001      // How fast to increase gain (smooth)
    private val agcMaxGain = 8.0         // Maximum amplification
    private val agcMinGain = 0.5         // Minimum amplification

    fun applyAGC(buffer: ByteArray, length: Int) {
        val samples = length / 2

        // Calculate RMS (root mean square) for this buffer
        var sumSquares = 0.0
        for (i in 0 until samples) {
            val rawSample = (buffer[i * 2].toInt() and 0xFF) or (buffer[i * 2 + 1].toInt() shl 8)
            val sample = if (rawSample > 32767) rawSample - 65536 else rawSample
            sumSquares += sample * sample
        }
        val rms = if (samples > 0) kotlin.math.sqrt(sumSquares / samples) else 0.0

        // Calculate desired gain to reach target RMS
        val desiredGain = if (rms > 0) agcTarget / rms else 1.0

        // Smooth gain changes (attack/release)
        agcCurrentGain = if (desiredGain < agcCurrentGain) {
            // Reduce gain quickly (attack) to prevent clipping
            agcCurrentGain + (desiredGain - agcCurrentGain) * agcAttack
        } else {
            // Increase gain slowly (release) for smooth transitions
            agcCurrentGain + (desiredGain - agcCurrentGain) * agcRelease
        }

        // Clamp gain to safe range
        agcCurrentGain = agcCurrentGain.coerceIn(agcMinGain, agcMaxGain)

        // Apply gain to buffer
        for (i in 0 until samples) {
            val rawSample = (buffer[i * 2].toInt() and 0xFF) or (buffer[i * 2 + 1].toInt() shl 8)
            val sample = if (rawSample > 32767) rawSample - 65536 else rawSample

            // Apply gain with soft clipping
            val amplified = (sample * agcCurrentGain).toInt()
            val clipped = amplified.coerceIn(-32768, 32767)
            val unsigned = if (clipped < 0) clipped + 65536 else clipped

            buffer[i * 2] = (unsigned and 0xFF).toByte()
            buffer[i * 2 + 1] = (unsigned shr 8).toByte()
        }
    }

    fun resetAGC() {
        agcCurrentGain = 1.0
    }


    // ==================== Pre-emphasis Filter ====================

    /**
     * Pre-emphasis filter - boosts high frequencies for better consonant clarity.
     *
     * Standard in telephony (ITU-T G.711). Improves recognition of:
     * - Consonants (s, t, k, p, f, etc.)
     * - Word boundaries
     * - Overall speech intelligibility
     *
     * Uses coefficient of 0.97 (industry standard)
     */
    private var preEmphasisPrevSample = 0

    fun applyPreEmphasis(buffer: ByteArray, length: Int, coefficient: Double = 0.97) {
        val samples = length / 2

        for (i in 0 until samples) {
            val rawSample = (buffer[i * 2].toInt() and 0xFF) or (buffer[i * 2 + 1].toInt() shl 8)
            val sample = if (rawSample > 32767) rawSample - 65536 else rawSample

            // y[n] = x[n] - α * x[n-1]
            val emphasized = sample - (coefficient * preEmphasisPrevSample).toInt()
            preEmphasisPrevSample = sample

            // Clamp to 16-bit range
            val clipped = emphasized.coerceIn(-32768, 32767)
            val unsigned = if (clipped < 0) clipped + 65536 else clipped

            buffer[i * 2] = (unsigned and 0xFF).toByte()
            buffer[i * 2 + 1] = (unsigned shr 8).toByte()
        }
    }

    fun resetPreEmphasis() {
        preEmphasisPrevSample = 0
    }


    // ==================== Noise Gate ====================

    /**
     * Noise Gate - aggressively suppresses audio below threshold.
     *
     * More aggressive than VAD - completely zeros out quiet sections:
     * - Breathing between words
     * - Constant background hum
     * - Keyboard typing
     * - Air conditioning
     *
     * Uses hysteresis (different open/close thresholds) to prevent flutter.
     */
    private var gateIsOpen = false
    private val gateOpenThreshold = 600    // Must exceed to open gate
    private val gateCloseThreshold = 300   // Must fall below to close gate
    private val gateAttackSamples = 10     // Samples to fade in
    private val gateReleaseSamples = 50    // Samples to fade out
    private var gateFadeCounter = 0

    fun applyNoiseGate(buffer: ByteArray, length: Int): Boolean {
        val samples = length / 2

        // Calculate average amplitude
        var sum = 0L
        for (i in 0 until samples) {
            val rawSample = (buffer[i * 2].toInt() and 0xFF) or (buffer[i * 2 + 1].toInt() shl 8)
            val sample = if (rawSample > 32767) rawSample - 65536 else rawSample
            sum += abs(sample).toLong()
        }
        val avgAmplitude = if (samples > 0) (sum / samples).toInt() else 0

        // Hysteresis: different thresholds for opening vs closing
        val shouldBeOpen = if (gateIsOpen) {
            avgAmplitude > gateCloseThreshold  // Stay open until below close threshold
        } else {
            avgAmplitude > gateOpenThreshold   // Must exceed open threshold to open
        }

        // Update gate state with fade in/out
        if (shouldBeOpen && !gateIsOpen) {
            // Opening gate - fade in
            gateIsOpen = true
            gateFadeCounter = 0
        } else if (!shouldBeOpen && gateIsOpen) {
            // Closing gate - fade out
            gateIsOpen = false
            gateFadeCounter = gateReleaseSamples
        }

        // Apply gate with fade
        var shouldZeroBuffer = false
        if (!gateIsOpen || gateFadeCounter > 0) {
            val fadeMultiplier = if (gateIsOpen) {
                // Fading in (attack)
                min(1.0, gateFadeCounter.toDouble() / gateAttackSamples)
            } else {
                // Fading out (release) or fully closed
                if (gateFadeCounter > 0) {
                    gateFadeCounter.toDouble() / gateReleaseSamples
                } else {
                    0.0 // Fully closed, zero the buffer
                }
            }

            if (fadeMultiplier == 0.0) {
                shouldZeroBuffer = true
            } else {
                // Apply fade
                for (i in 0 until samples) {
                    val rawSample = (buffer[i * 2].toInt() and 0xFF) or (buffer[i * 2 + 1].toInt() shl 8)
                    val sample = if (rawSample > 32767) rawSample - 65536 else rawSample

                    val faded = (sample * fadeMultiplier).toInt()
                    val unsigned = if (faded < 0) faded + 65536 else faded

                    buffer[i * 2] = (unsigned and 0xFF).toByte()
                    buffer[i * 2 + 1] = (unsigned shr 8).toByte()
                }
            }

            // Update fade counter
            if (gateIsOpen && gateFadeCounter < gateAttackSamples) {
                gateFadeCounter++
            } else if (!gateIsOpen && gateFadeCounter > 0) {
                gateFadeCounter--
            }
        }

        return shouldZeroBuffer
    }

    fun resetNoiseGate() {
        gateIsOpen = false
        gateFadeCounter = 0
    }


    // ==================== Click/Pop Removal (Median Filter) ====================

    /**
     * Median filter - removes impulse noise (clicks, pops, taps).
     *
     * Detects sudden amplitude spikes and replaces with median of neighbors.
     * Effective for:
     * - Phone taps/touches
     * - Mouth clicks
     * - Pocket rustling
     * - Key jingles
     */
    private val clickWindowSize = 5  // Samples to look at (must be odd)
    private val clickWindow = IntArray(clickWindowSize)
    private var clickWindowIndex = 0

    fun removeClicks(buffer: ByteArray, length: Int) {
        val samples = length / 2

        for (i in 0 until samples) {
            val rawSample = (buffer[i * 2].toInt() and 0xFF) or (buffer[i * 2 + 1].toInt() shl 8)
            val sample = if (rawSample > 32767) rawSample - 65536 else rawSample

            // Add to circular buffer
            clickWindow[clickWindowIndex] = sample
            clickWindowIndex = (clickWindowIndex + 1) % clickWindowSize

            // Calculate median of window
            val sorted = clickWindow.sorted()
            val median = sorted[clickWindowSize / 2]

            // Detect spike: if current sample is very different from median
            val diff = abs(sample - median)
            val threshold = 5000  // Spike threshold

            val output = if (diff > threshold) {
                // Likely a click/pop - use median instead
                median
            } else {
                // Normal audio - pass through
                sample
            }

            val unsigned = if (output < 0) output + 65536 else output
            buffer[i * 2] = (unsigned and 0xFF).toByte()
            buffer[i * 2 + 1] = (unsigned shr 8).toByte()
        }
    }

    fun resetClickRemoval() {
        clickWindow.fill(0)
        clickWindowIndex = 0
    }


    // ==================== Master Reset ====================

    /**
     * Reset all enhancement states (call when starting new recording)
     */
    fun resetAll() {
        resetAGC()
        resetPreEmphasis()
        resetNoiseGate()
        resetClickRemoval()
    }
}
